src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -265,43 +267,90 @@ class DeepDiveLogParser extends JavaTokenParsers {
-      case (q ~ supervision) => ExtractionRule(q, supervision)
-    }
+  def cqBody: Parser[Body] = cond | quantifiedBody | atom
+
+  def cqConjunctiveBody: Parser[List[Body]] = rep1sep(cqBody, ",")
+
+  def cqHeadTerms = "(" ~> rep1sep(expr, ",") <~ ")"
+
+  def conjunctiveQuery : Parser[ConjunctiveQuery] =
+    cqHeadTerms ~ opt("*") ~ opt("|" ~> decimalNumber) ~ ":-" ~ rep1sep(cqConjunctiveBody, ";") ^^ {
+      case (head ~ isDistinct ~ limit ~ ":-" ~ disjunctiveBodies) =>
+        ConjunctiveQuery(head, disjunctiveBodies, isDistinct != None, limit map (_.toInt))
+  }
 
   def functionCallRule : Parser[FunctionCallRule] =
-    ( relationName ~ ":-" ~ "!"
-    ~ functionName ~ "(" ~ relationName ~ ")"
-    ) ^^ {
-      case (out ~ ":-" ~ "!" ~ fn ~ "(" ~ in ~ ")") =>
-        FunctionCallRule(in, out, fn)
+    relationName ~ "+=" ~ functionName ~ conjunctiveQuery ^^ {
+      case (out ~ _ ~ func ~ cq) => FunctionCallRule(out, func, cq)
     }
 
-  def factorWeight = "weight" ~> "=" ~> rep1sep(expr, ",") ^^ { FactorWeight(_) }
+  def supervision = "@label" ~> "(" ~> variableName <~ ")"
+  def extractionRule =
+    opt(supervision) ~ relationName ~ conjunctiveQuery ^^ {
+      case (sup ~ head ~ cq) => ExtractionRule(head, cq, sup)
+  }
+
+  def factorWeight = "@weight" ~> "(" ~> rep1sep(expr, ",") <~ ")" ^^ { FactorWeight(_) }
+  def inferenceMode = "@mode" ~> "(" ~> inferenceModeType <~ ")"
 
-  def supervision = "label" ~> "=" ~> variableName
+  // factor functions
+  def headAtom = relationName ~ ("(" ~> rep1sep(expr, ",") <~ ")") ^^ {
+    case (r ~ expressions) => HeadAtom(r, expressions)
+  }
 
-  def factorFunctionName = "Imply" | "And" | "Equal" | "Or" | "Multinomial" | "Linear" | "Ratio"
-  def factorFunction = "function" ~> "=" ~> factorFunctionName
+  def implyHeadAtoms = rep1sep(headAtom, ",") ~ "=>" ~ headAtom ^^ {
+    case (a ~ _ ~ b) => a :+ b
+  }
 
-  def inferenceRule : Parser[InferenceRule] =
-    ( conjunctiveQuery ~ factorWeight ~ opt(factorFunction) ~ opt(inferenceMode)
-    ) ^^ {
-      case (q ~ weight ~ function ~ mode) =>
-        InferenceRule(q, weight, function, mode.getOrElse(null))
+  def inferenceRuleHead =
+  ( implyHeadAtoms ^^ {
+      InferenceRuleHead(FactorFunction.Imply, _)
     }
+  | headAtom ~ "=" ~ rep1sep(headAtom, "=") ^^ { case (a ~ _ ~ b) =>
+      InferenceRuleHead(FactorFunction.Equal, a +: b)
+    }
+  | headAtom ~ "^" ~ rep1sep(headAtom, "^") ^^ { case (a ~ _ ~ b) =>
+      InferenceRuleHead(FactorFunction.And, a +: b)
+    }
+  | headAtom ~ "v" ~ rep1sep(headAtom, "v") ^^ { case (a ~ _ ~ b) =>
+      InferenceRuleHead(FactorFunction.Or, a +: b)
+    }
+  | "Multinomial" ~> "(" ~> rep1sep(headAtom, ",") <~ ")" ^^ {
+      InferenceRuleHead(FactorFunction.Multinomial, _)
+    }
+  | "@semantics(linear)" ~> implyHeadAtoms ^^ {
+      InferenceRuleHead(FactorFunction.Linear, _)
+    }
+  | "@semantics(ratio)" ~> implyHeadAtoms ^^ {
+      InferenceRuleHead(FactorFunction.Ratio, _)
+    }
```

Comment:
Can you move these imply cases up right below the first one with logical semantics?  I feel they should be grouped together to also make it clear why `implyHeadAtoms` is there.